packages/services/src/view-service/transaction-planner/index.ts

```diff
@@ -7,22 +7,29 @@ import { TransactionPlannerRequest } from '@buf/penumbra-zone_penumbra.bufbuild_
 import { fvkCtx } from '../../ctx/full-viewing-key.js';
 import { extractAltFee } from '../fees.js';
 import { assertTransactionSource } from './assert-transaction-source.js';
+import { AddressIndex } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/keys/v1/keys_pb.js';
+import { toPlainMessage } from '@bufbuild/protobuf';
 
 export const transactionPlanner: Impl['transactionPlanner'] = async (req, ctx) => {
   assertValidRequest(req);
 
   const services = await ctx.values.get(servicesCtx)();
   const { indexedDb } = await services.getWalletServices();
 
-  // Query IndexedDB directly to check for the existence of staking token
-  const nativeToken = await indexedDb.hasTokenBalance(req.source!, indexedDb.stakingTokenAssetId);
+  const noNativeTokenAvailable = !(await indexedDb.accountHasSpendableAsset(
+    toPlainMessage(req.source ?? new AddressIndex()),
+    indexedDb.stakingTokenAssetId,
+  ));
 
+  const planningSwapClaim = req.swapClaims.length > 0;
   // Check if we should use the native token or extract an alternate gas fee token.
   // Special cased for swap claims as gas fee needs to match the claimFee on the corresponding swap.
-  const needsAltFeeToken = !nativeToken || req.swapClaims.length > 0;
-  const gasFeeToken = needsAltFeeToken
-    ? await extractAltFee(req, indexedDb)
-    : indexedDb.stakingTokenAssetId;
+  console.log('alt fee reasons', noNativeTokenAvailable, planningSwapClaim);
+
+  const gasFeeToken =
+    noNativeTokenAvailable || planningSwapClaim
```

Comment:
I actually prefer keeping `needsAltFeeToken`, so we can tack on more invariants later.